### ISSUE TITLE
Improving multiline further

### ISSUE THREAD

**Issue by trygveaa** [weechat/weechat#1498]
## Feature description

Thanks for merging the multiline PR and fixing the input display! With this, scripts can have proper multiline support without using hacks as long as the user either has the `multiline.pl` script or is using the kitty or alacritty terminal emulators and set some options. However,  I have some thoughts on improving multiline further.

There are two ways to input multiline lines. Either by inserting a newline manually, or by pasing multiple lines.

### Inserting newline manually

This can be done by binding a key to `/input insert \n`. It would be nice to have a default key for this. Personally, I use alt+enter which is nice because it is similar to shift+enter used in many other applications (but unfortunately can't be used in terminals). However, alt+enter makes the application fullscreen in xterm (and on windows), so maybe some other key is better.

I see that you only applied multiline display for buffers that has `input_multiline` set. I think it's practical to be able to edit multiple lines even if it will be sent as multiple lines to the buffer, so I think it would be better if it was done for all buffers. If so, the binding is useful for all buffers. If you don't want to do this for all buffers, the binding should probably be a noop when `input_multiline` isn't set.

### Pasting

The kitty and alacritty terminal emulators sends multiple lines separated by newline characters, while other terminals I've tested separates them by carriage return. This means that to make paste work for multiple lines with in kitty/alacritty, you just have to bind `ctrl-J` to `/input insert \n`.

However, since this doesn't work for other terminals, it would be nice if bracketed paste could be used so that carriage return doesn't send the line when a paste occurs. Again, I think this would be useful even for buffers without `input_multiline` set, so you are able to edit the pasted lines.

If this is done, the paste confirmation isn't really that useful anymore imo, since a message isn't sent until you press enter, so it could be disabled by default, but this is subjective.

Additionally, it would be nice to trim the last newline after a paste, because if you have copied multiple full lines, you end up with a blank line at the end of the input, which is usually not what you want.

### Input bar size

To view multiple lines in the input bar you have to adjust `weechat.bar.input.size`. The most useful is to set it to 0, so it remains the same size when you only have one line, but adjusts dynamically when you insert multiple. I don't really see a reason to have it set to 1, so in my opinion, the default could just be changed.

**Comment by trygveaa** [comment]
I also noticed some additional issues:
- If you try to insert multiple consecutive lines, i.e. blank lines, they are not shown in the input. However, when you send the message the script does receive them, so it's just a display issue.
- If you try to paste something which contains blank lines, they are removed. In this case they are completely removed, the script does not receive them.

**Comment by trygveaa** [comment]
> If you try to paste something which contains blank lines, they are removed. In this case they are completely removed, the script does not receive them.

This seems to be intentional? It's caused by this: https://github.com/weechat/weechat/blob/867e07aa18598ed04abddc342f8e73a3ac61c0bc/src/gui/curses/gui-curses-key.c#L550-L557

**Comment by trygveaa** [comment]
> If you try to insert multiple consecutive lines, i.e. blank lines, they are not shown in the input. However, when you send the message the script does receive them, so it's just a display issue.

This is caused by `WEECHAT_STRING_SPLIT_COLLAPSE_SEPS` here:
https://github.com/weechat/weechat/blob/867e07aa18598ed04abddc342f8e73a3ac61c0bc/src/gui/curses/gui-curses-bar-window.c#L490-L494

**Comment by flashcode** [comment]
> > If you try to insert multiple consecutive lines, i.e. blank lines, they are not shown in the input. However, when you send the message the script does receive them, so it's just a display issue.
> 
> This is caused by `WEECHAT_STRING_SPLIT_COLLAPSE_SEPS` here:
> 
> https://github.com/weechat/weechat/blob/867e07aa18598ed04abddc342f8e73a3ac61c0bc/src/gui/curses/gui-curses-bar-window.c#L490-L494

This is a regression when I introduced flags in function string_split. In the commit I changed a lot of calls to this function, I'll check if there are not other regressions in code.

**Comment by hasufell** [comment]
> This means that to make paste work for multiple lines with in kitty/alacritty, you just have to bind ctrl-J to /input insert \n.

I tried this with `multiline.pl` and alacritty and it didn't work. I get no newlines.

**Comment by trygveaa** [comment]
> I tried this with `multiline.pl` and alacritty and it didn't work. I get no newlines.

That's because `multiline.pl` overwrites the `ctrl-J` binding. Either unload the script or run `/set plugins.var.perl.multiline.weechat_paste_fix "off"`. You have to set the binding again after doing this.

I've also opened PR #1502 to make pasting work in all terminals by default, without having to use `multiline.pl` or set a binding.

**Comment by grassdionera** [comment]
Hello,

today i tried to get the multiline plugin to work with weechat 3.3, wee-slack plugin with alacritty terminal. After hours of trying i realized that the option `paste_bracketed` was set to on. after setting it to off in the weechat.conf,
the paste insert with `\n` or `\r\n` is working again.

**Comment by trygveaa** [comment]
@grassdionera: Setting `paste_bracketed` to off is not really what you want to do since alacritty supports bracketed paste, and disabling it means weechat having to detect multiple lines using timers instead, which is less reliable. I'm guessing your issue is caused by alacritty sending LF instead of CR between each line when you paste and `multiline.pl` overriding ctrl-J (LF). If so  you can fix the issue by running these commands:

```
/set plugins.var.perl.multiline.weechat_paste_fix "off"
/key bind ctrl-J /input magic_enter
```

@flashcode: Things like this is why it would be nice to get my PRs to improve this merged :)

**Comment by grassdionera** [comment]
@trygveaa many thanks for this. It's working with paste_fix off and rebinding ctrl-J and paste_bracket on. 
@flashcode i would also like to see the PRs merged because the multiline.pl is a really old script and in future i would like to avoid it. Thanks

**Comment by flashcode** [comment]
@trygveaa: FYI I scheduled work and merge of PR #1502 for the version 3.5, so a merge can occur after the release of 3.4 (scheduled on Dec 18th, 2021).

**Comment by trygveaa** [comment]
@flashcode: Great, thanks! Would be nice if you could look at #1503 and #1504 at the same time. They are related, and are pretty simple, so I think they should be easier to review than #1502. I can fix the merge conflicts in #1503.

**Comment by flashcode** [comment]
All done in v4.0.0 :)
